packages/core/src/migration/index.ts

```diff
@@ -0,0 +1,142 @@
+
+const isLogtoConfigsTableExists = async (pool: DatabasePool) => {
+  const [result] = await pool.anyFirst<[boolean]>(sql`
+    select exists (
+      select from 
+        pg_tables
+      where 
+        tablename = ${logtoConfigsTable}
+      );
+  `);
+
+  return result;
+};
+
+const getCurrentDatabaseVersion = async (pool: DatabasePool) => {
+  if (!(await isLogtoConfigsTableExists(pool))) {
+    return null;
+  }
+
+  const query = await pool.maybeOne<LogtoConfig>(
+    sql`select * from ${sql.identifier([logtoConfigsTable])} where key=${databaseVersionKey}`
+  );
+
+  if (!query) {
+    return null;
+  }
+
+  try {
+    const databaseVersion = databaseVersionGuard.parse(query.value);
+
+    return databaseVersion.version;
+  } catch {
+    return null;
+  }
+};
+
+const createLogtoConfigsTable = async (pool: DatabasePool) => {
+  const tableQuery = await readFile(logtoConfigsTableFilePath, 'utf8');
+  await pool.query(sql`${raw(tableQuery)}`);
+};
+
+const updateDatabaseVersion = async (pool: DatabasePool, version: string) => {
+  if (!(await isLogtoConfigsTableExists(pool))) {
+    await createLogtoConfigsTable(pool);
+  }
+
+  const value: DatabaseVersion = {
+    version,
+    updatedAt: new Date().toISOString(),
+  };
+
+  await pool.query(
+    sql`
+      insert into ${sql.identifier([logtoConfigsTable])} (key, value) 
+        values (${databaseVersionKey}, ${JSON.stringify(value)})
+        on conflict (key) do update set value=excluded.value
+    `
+  );
+};
+
+const getMigrationFiles = async () => {
+  const directory = await readdir(migrationFilesDirectory);
+  const files = directory.filter(
+    (file) => migrationFileNameRegex.test(file) && getVersionFromFileName(file) !== 'next'
+  );
+
+  return files;
+};
+
+export const getPendingMigrations = async (pool: DatabasePool) => {
+  const databaseVersion = await getCurrentDatabaseVersion(pool);
+  const files = await getMigrationFiles();
+
+  if (!databaseVersion) {
+    return files;
+  }
+
+  return files
+    .filter((file) => compareVersion(getVersionFromFileName(file), databaseVersion) > 0)
+    .slice()
+    .sort((file1, file2) =>
+      compareVersion(getVersionFromFileName(file1), getVersionFromFileName(file2))
+    );
+};
+
+const importMigration = async (file: string): Promise<MigrationScript> => {
+  // eslint-disable-next-line no-restricted-syntax
+  const module = (await import(
+    path.join(migrationFilesDirectory, file).replace('node_modules/', '')
+  )) as MigrationScript;
+
+  return module;
+};
+
+const runMigration = async (pool: DatabasePool, file: string) => {
+  const { up } = await importMigration(file);
+
+  try {
+    await up(pool);
+    await updateDatabaseVersion(pool, getVersionFromFileName(file));
+    console.log(`${chalk.blue('[migration]')} run ${file} succeeded.`);
+  } catch (error: unknown) {
+    if (error instanceof Error) {
+      console.log(`${chalk.red('[migration]')} run ${file} failed: ${error.message}.`);
```

Comment:
Better to split the errors of the migration script and dbVersionBumping script?  If dbVersionBumping throws errors, technically migration did run sussefully.

Reply:
updated